abci/client/committing_client.go

```diff
@@ -160,6 +160,13 @@ func (app *committingClient) InitChainAsync(req types.RequestInitChain) *ReqRes
 }
 
 func (app *committingClient) BeginBlockAsync(req types.RequestBeginBlock) *ReqRes {
+	if len(req.Header.AppHash) > 0 && !app.mtx.IsInitialized() {
+		// We already have some state, so mark as initialized.
+		app.mtx.Lock()
+		app.mtx.Initialize()
+		app.mtx.Unlock()
```

Comment:
Consider making this `defer app.mtx.Unlock()` and move it up right after the `app.mtx.Lock()`, as this is more idiomatic and handles the case where `app.mtx.Initialize()` misbehaves.  Here and below.

Reply:
Nevermind - I was imagining this was followed by a `return` statement, but it isn't.